packages/contracts/test/contracts/discovery/Batch.ts

```diff
@@ -0,0 +1,276 @@
+import { ethers } from "hardhat";
+import { expect } from "chai";
+
+import type { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+import {
+  Batch,
+  Batch__factory,
+  Project,
+  Project__factory,
+} from "../../../src/types";
+
+import { goToTime, currentTimestamp } from "../../timeHelpers";
+
+const { parseUnits } = ethers.utils;
+
+describe("Batch", () => {
+  let owner: SignerWithAddress;
+  let alice: SignerWithAddress;
+  let bob: SignerWithAddress;
+  let fakeToken: SignerWithAddress;
+
+  let batch: Batch;
+  let fakeProject: Project;
+  let anotherFakeProject: Project;
+  let oneDay: number;
+  let votingStart: number;
+  let votingEnd: number;
+
+  beforeEach(async () => {
+    [owner, alice, bob, fakeToken] = await ethers.getSigners();
+    oneDay = 60 * 60 * 24;
+    votingStart = (await currentTimestamp()) + oneDay;
+    votingEnd = votingStart + oneDay * 10;
+
+    fakeProject = await new Project__factory(owner).deploy(
+      "A project",
+      fakeToken.address,
+      1000,
+      10
+    );
+
+    anotherFakeProject = await new Project__factory(owner).deploy(
+      "Another project",
+      fakeToken.address,
+      1000,
+      10
+    );
+
+    batch = await new Batch__factory(owner).deploy(
+      [fakeProject.address, anotherFakeProject.address],
+      2
+    );
+
+    await batch.setVotingPeriod(votingStart, votingEnd);
+    await goToTime(votingStart);
+  });
+
+  describe("constructor", async () => {
+    it("should set the correct values", async () => {
+      expect(await batch.projects(0)).to.eq(fakeProject.address);
+      expect(await batch.projects(1)).to.eq(anotherFakeProject.address);
+      expect(await batch.slotCount()).to.eq(2);
+    });
+
+    it("fails with an empty list of projects", async () => {
+      await expect(new Batch__factory(owner).deploy([], 0)).to.be.revertedWith(
+        "projects must not be empty"
+      );
+    });
+
+    it("fails with an address that does not implement IProject", async () => {
+      await expect(
+        new Batch__factory(owner).deploy([fakeToken.address], 1)
+      ).to.be.revertedWith("project must be an IProject");
+    });
+
+    // it("fails if one of the projects is not whitelisted", async () => {
+    //   await expect(
+    //     new Batch__factory(owner).deploy([fakeProject.address], 1)
+    //   ).to.be.revertedWith("project must be whitelisted");
+    // });
```

Comment:
is this meant to be commented?

Reply:
I can make it pending instead. Don't know what a whitelisted project is yet.

Reply:
It's a project that answers true to `IProject(project).isReadyForListing()`